File: services/elevation.py
"""
Terrain elevation service using Open-Meteo Elevation API.
Free, no API key required: api.open-meteo.com/v1/elevation
"""
import requests

ELEVATION_API_URL = "https://api.open-meteo.com/v1/elevation"
_cache = {}


def get_elevations(coordinates):
    """Get ground elevation for a list of (lat, lng) tuples.

    Args:
        coordinates: list of (lat, lng) tuples

    Returns:
        list of elevation values in metres (or None if API fails)
    """
    if not coordinates:
        return []

    # Check cache first
    uncached = []
    uncached_indices = []
    results = [None] * len(coordinates)

    for i, (lat, lng) in enumerate(coordinates):
        key = (round(lat, 5), round(lng, 5))
        if key in _cache:
            results[i] = _cache[key]
        else:
            uncached.append((lat, lng))
            uncached_indices.append(i)

    if not uncached:
        return results

    # Batch API call (max 100 at a time)
    batch_size = 100
    for start in range(0, len(uncached), batch_size):
        batch = uncached[start:start + batch_size]
        lats = ",".join(f"{lat:.5f}" for lat, _ in batch)
        lngs = ",".join(f"{lng:.5f}" for _, lng in batch)

        try:
            resp = requests.get(
                ELEVATION_API_URL,
                params={"latitude": lats, "longitude": lngs},
                timeout=10,
            )
            resp.raise_for_status()
            data = resp.json()
            elevations = data.get("elevation", [])

            for j, elev in enumerate(elevations):
                idx = uncached_indices[start + j]
                lat, lng = uncached[start + j]
                key = (round(lat, 5), round(lng, 5))
                _cache[key] = elev
                results[idx] = elev
        except Exception:
            # Fallback: return 0 for failed lookups
            for j in range(len(batch)):
                idx = uncached_indices[start + j]
                if results[idx] is None:
                    results[idx] = 0.0

    return results


def get_elevation(lat, lng):
    """Get elevation for a single point."""
    result = get_elevations([(lat, lng)])
    return result[0] if result else 0.0


def get_path_elevations(start_lat, start_lng, end_lat, end_lng, num_samples=10):
    """Get elevations along a path between two points.

    Returns list of dicts with lat, lng, elevation, distance_m.
    """
    coords = []
    for i in range(num_samples):
        t = i / max(num_samples - 1, 1)
        lat = start_lat + t * (end_lat - start_lat)
        lng = start_lng + t * (end_lng - start_lng)
        coords.append((lat, lng))

    elevations = get_elevations(coords)

    results = []
    total_dist = 0
    for i, ((lat, lng), elev) in enumerate(zip(coords, elevations)):
        if i > 0:
            from math import radians, sin, cos, sqrt, atan2
            dlat = radians(lat - coords[i - 1][0])
            dlng = radians(lng - coords[i - 1][1])
            a = sin(dlat / 2) ** 2 + cos(radians(coords[i - 1][0])) * cos(radians(lat)) * sin(dlng / 2) ** 2
            total_dist += 6371000 * 2 * atan2(sqrt(a), sqrt(1 - a))

        results.append({
            "lat": lat,
            "lng": lng,
            "elevation": elev,
            "distance_m": round(total_dist, 1),
        })

    return results


def get_waypoint_elevations(waypoints):
    """Get elevations for a list of waypoint dicts. Returns enriched list."""
    if not waypoints:
        return []

    coords = [(w["lat"], w["lng"]) for w in waypoints]
    elevations = get_elevations(coords)

    result = []
    for wp, elev in zip(waypoints, elevations):
        enriched = dict(wp)
        enriched["ground_elevation_m"] = elev
        enriched["amsl_m"] = (elev or 0) + wp.get("altitude_m", 30)
        enriched["agl_m"] = wp.get("altitude_m", 30)
        result.append(enriched)

    return result


def clear_cache():
    """Clear the elevation cache."""
    _cache.clear()

File: php/app/Services/Elevation.php
<?php

namespace App\Services;

/**
 * Terrain elevation service using Open-Meteo Elevation API.
 * Free, no API key required.
 * Ported from services/elevation.py.
 */
class Elevation
{
    private const API_URL = 'https://api.open-meteo.com/v1/elevation';
    private static array $cache = [];

    /**
     * Get ground elevation for a list of [lat, lng] pairs.
     * @return float[] elevation values in metres
     */
    public static function getElevations(array $coordinates): array
    {
        if (empty($coordinates)) {
            return [];
        }

        $uncached = [];
        $uncachedIndices = [];
        $results = array_fill(0, count($coordinates), null);

        foreach ($coordinates as $i => [$lat, $lng]) {
            $key = round($lat, 5) . ',' . round($lng, 5);
            if (isset(self::$cache[$key])) {
                $results[$i] = self::$cache[$key];
            } else {
                $uncached[] = [$lat, $lng];
                $uncachedIndices[] = $i;
            }
        }

        if (empty($uncached)) {
            return $results;
        }

        // Batch API call (max 100 at a time)
        $batchSize = 100;
        for ($start = 0; $start < count($uncached); $start += $batchSize) {
            $batch = array_slice($uncached, $start, $batchSize);
            $lats = implode(',', array_map(fn($c) => sprintf('%.5f', $c[0]), $batch));
            $lngs = implode(',', array_map(fn($c) => sprintf('%.5f', $c[1]), $batch));

            try {
                $url = self::API_URL . '?' . http_build_query([
                    'latitude' => $lats,
                    'longitude' => $lngs,
                ]);

                $ctx = stream_context_create(['http' => ['timeout' => 10]]);
                $response = @file_get_contents($url, false, $ctx);

                if ($response !== false) {
                    $data = json_decode($response, true);
                    if (!is_array($data)) {
                        log_message('warning', 'Elevation API returned invalid JSON');
                        continue;
                    }
                    $elevations = $data['elevation'] ?? [];

                    foreach ($elevations as $j => $elev) {
                        $idx = $uncachedIndices[$start + $j];
                        [$lat, $lng] = $uncached[$start + $j];
                        $key = round($lat, 5) . ',' . round($lng, 5);
                        self::$cache[$key] = $elev;
                        $results[$idx] = $elev;
                    }
                }
            } catch (\Exception $e) {
                // Fallback: 0 for failed lookups
            }

            // Fill any remaining nulls with 0
            for ($j = 0; $j < count($batch); $j++) {
                $idx = $uncachedIndices[$start + $j];
                if ($results[$idx] === null) {
                    $results[$idx] = 0.0;
                }
            }
        }

        return $results;
    }

    public static function getElevation(float $lat, float $lng): float
    {
        $result = self::getElevations([[$lat, $lng]]);
        return $result[0] ?? 0.0;
    }

    /**
     * Get elevations along a path between two points.
     */
    public static function getPathElevations(
        float $startLat, float $startLng, float $endLat, float $endLng,
        int $numSamples = 10
    ): array {
        $coords = [];
        for ($i = 0; $i < $numSamples; $i++) {
            $t = $i / max($numSamples - 1, 1);
            $coords[] = [
                $startLat + $t * ($endLat - $startLat),
                $startLng + $t * ($endLng - $startLng),
            ];
        }

        $elevations = self::getElevations($coords);

        $results = [];
        $totalDist = 0;
        foreach ($coords as $i => [$lat, $lng]) {
            if ($i > 0) {
                $totalDist += GeoUtils::haversine(
                    $coords[$i - 1][0], $coords[$i - 1][1], $lat, $lng
                );
            }
            $results[] = [
                'lat' => $lat, 'lng' => $lng,
                'elevation' => $elevations[$i],
                'distance_m' => round($totalDist, 1),
            ];
        }

        return $results;
    }

    /**
     * Enrich waypoints with ground elevation data.
     */
    public static function getWaypointElevations(array $waypoints): array
    {
        if (empty($waypoints)) {
            return [];
        }

        $coords = array_map(fn($w) => [$w['lat'], $w['lng']], $waypoints);
        $elevations = self::getElevations($coords);

        $result = [];
        foreach ($waypoints as $i => $wp) {
            $enriched = $wp;
            $enriched['ground_elevation_m'] = $elevations[$i];
            $enriched['amsl_m'] = ($elevations[$i] ?? 0) + ($wp['altitude_m'] ?? 30);
            $enriched['agl_m'] = $wp['altitude_m'] ?? 30;
            $result[] = $enriched;
        }

        return $result;
    }

    public static function clearCache(): void
    {
        self::$cache = [];
    }
}
